File: skills/salesforce-anonymous-apex-debugging/scripts/extract_log_snippet.js
#!/usr/bin/env node

/**
 * Extract a focused snippet from a Salesforce debug log.
 *
 * Usage:
 *   node extract_log_snippet.js --log /path/to/log.txt --token MYBUG --context 80
 *
 * What it does:
 *   - Prints lines around the first occurrence of the token.
 *   - Also prints the first occurrence of EXCEPTION_THROWN or FATAL_ERROR if present.
 *
 * Notes:
 *   - This is a convenience tool for humans and agents.
 *   - It does not parse log structure beyond simple string matching.
 */

const fs = require("node:fs");

function parseArgs(argv) {
  const args = { context: 80 };

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--log") {
      args.log = argv[i + 1];
      i += 1;
    } else if (arg === "--token") {
      args.token = argv[i + 1];
      i += 1;
    } else if (arg === "--context") {
      const parsed = Number.parseInt(argv[i + 1], 10);
      if (Number.isNaN(parsed) || parsed < 0) {
        throw new Error("--context must be a non-negative integer");
      }
      args.context = parsed;
      i += 1;
    } else if (arg === "--help" || arg === "-h") {
      printUsageAndExit(0);
    } else {
      throw new Error(`Unknown argument: ${arg}`);
    }
  }

  if (!args.log || !args.token) {
    printUsageAndExit(1);
  }

  return args;
}

function printUsageAndExit(code) {
  const usage = [
    "Usage:",
    "  node extract_log_snippet.js --log /path/to/log.txt --token MYBUG --context 80",
    "",
    "Required arguments:",
    "  --log      Path to the debug log text file",
    "  --token    Token to locate, example: MYBUG",
    "",
    "Optional arguments:",
    "  --context  Lines before and after the match (default: 80)",
  ].join("\n");
  process.stdout.write(`${usage}\n`);
  process.exit(code);
}

function findFirstIndex(lines, needle) {
  for (let i = 0; i < lines.length; i += 1) {
    if (lines[i].includes(needle)) {
      return i;
    }
  }
  return -1;
}

function printWindow(lines, center, context) {
  const start = Math.max(0, center - context);
  const end = Math.min(lines.length, center + context + 1);

  for (let i = start; i < end; i += 1) {
    const prefix = String(i + 1).padStart(6, " ");
    const line = lines[i].replace(/\r?\n$/, "");
    process.stdout.write(`${prefix}: ${line}\n`);
  }
}

function main() {
  const args = parseArgs(process.argv.slice(2));

  if (!fs.existsSync(args.log)) {
    process.stderr.write(`Log file not found: ${args.log}\n`);
    process.exit(1);
  }

  const lines = fs.readFileSync(args.log, "utf8").split(/\r?\n/);

  process.stdout.write("\n=== Token window ===\n");
  const tokenIndex = findFirstIndex(lines, args.token);
  if (tokenIndex === -1) {
    process.stdout.write(`Token not found: ${args.token}\n`);
  } else {
    printWindow(lines, tokenIndex, args.context);
  }

  process.stdout.write("\n=== Exception window ===\n");
  const exceptionThrownIndex = findFirstIndex(lines, "EXCEPTION_THROWN");
  const fatalErrorIndex = findFirstIndex(lines, "FATAL_ERROR");
  const systemIndex = findFirstIndex(lines, "System.");
  const exceptionIndex =
    exceptionThrownIndex !== -1
      ? exceptionThrownIndex
      : fatalErrorIndex !== -1
        ? fatalErrorIndex
        : systemIndex;

  if (exceptionIndex === -1) {
    process.stdout.write("No EXCEPTION_THROWN or FATAL_ERROR found\n");
  } else {
    printWindow(lines, exceptionIndex, args.context);
  }
}

main();

File: skills/salesforce-anonymous-apex-debugging/scripts/extract_log_snippet.py
"""Extract a focused snippet from a Salesforce debug log.

Usage:
  python extract_log_snippet.py --log /path/to/log.txt --token MYBUG --context 80

What it does:
  - Prints lines around the first occurrence of the token.
  - Also prints the first occurrence of EXCEPTION_THROWN or FATAL_ERROR if present.

Notes:
  - This is a convenience tool for humans and agents.
  - It does not parse log structure beyond simple string matching.
"""

from __future__ import annotations

import argparse
from pathlib import Path


def _find_first_index(lines: list[str], needle: str) -> int | None:
    for i, line in enumerate(lines):
        if needle in line:
            return i
    return None


def _print_window(lines: list[str], center: int, context: int) -> None:
    start = max(0, center - context)
    end = min(len(lines), center + context + 1)
    for i in range(start, end):
        prefix = f"{i+1:>6}: "
        print(prefix + lines[i].rstrip("\n"))


def main() -> None:
    p = argparse.ArgumentParser()
    p.add_argument("--log", required=True, help="Path to the debug log text file")
    p.add_argument("--token", required=True, help="Token to locate, example: MYBUG")
    p.add_argument("--context", type=int, default=80, help="Lines before and after the match")
    args = p.parse_args()

    path = Path(args.log)
    if not path.exists():
        raise SystemExit(f"Log file not found: {path}")

    lines = path.read_text(errors="replace").splitlines(True)

    print("\n=== Token window ===")
    token_index = _find_first_index(lines, args.token)
    if token_index is None:
        print(f"Token not found: {args.token}")
    else:
        _print_window(lines, token_index, args.context)

    print("\n=== Exception window ===")
    ex_index = (
        _find_first_index(lines, "EXCEPTION_THROWN")
        or _find_first_index(lines, "FATAL_ERROR")
        or _find_first_index(lines, "System.")
    )
    if ex_index is None:
        print("No EXCEPTION_THROWN or FATAL_ERROR found")
    else:
        _print_window(lines, ex_index, args.context)


if __name__ == "__main__":
    main()
